프로젝트 관리 앱 [STEP 2-2] Judy

Judy-999 opened the pull request:
# 프로젝트 관리 앱 Step 2-2
안녕하세요 라자냐 🤗
프로젝트 관리 앱 Step 2-2 PR 보냅니다!
Rx의 기능과 MVVM 구조를 잘 적용하고 있는지 봐주시면 감사하겠습니다 🙇‍♀️
<br>

## 코드 설명 👩🏻‍💻
Step 2-2에서 추가된 코드만 적었습니다 🙂

**Model**
- `WorkState`: 일의 진행 상태를 표현하는 열거형 (todo, doing, done)
- `Work` : 일의 고유한 값인 **id**와 `WorkState` 타입의 **state** 프로퍼티 추가
<br>

**View**
- `ChangeWorkStateViewController` : 할 일의 진행 상태를 변경하는 뷰컨트롤러
<br>

**ViewModel**
- `works` : 전체 할 일을 가진 `BehaviorRelay`
- `selectWork` : 인덱스(Int)와 상태로 `Work`를 찾아 반환하는 메서드
- `editWork` : 할 일의 내용을 변경하는 메서드
- `deleteWork` : 할 일을 삭제하는 메서드
- `chnageWorkState` : 특정 할 일의 상태를 변경하는 메서드
<br>


## 구현 화면 📱

| 새로운 할 일 생성 | 할 일 수정 | 
| :--------: | :--------: |
| ![](https://i.imgur.com/xwAX6GG.gif) |![](https://i.imgur.com/flJPZiB.gif)|

| Edit을 누르면 편집 가능 | 할 일 삭제 | 
| :--------: | :--------: |
| ![](https://i.imgur.com/WeY4FIJ.gif)|  ![](https://i.imgur.com/1LacwO1.gif) |

| 할 일 상태 변경 | 
| :--------: | 
|![](https://i.imgur.com/hug85LI.gif) |  

<br>

## 고민한 내용 🤔
### 1. 할 일 리스트를 하나만 가지도록 변경
이전에는 할 일의 상태에 따라 데이터를 처리하고 싶어 뷰모델에 아래와 같이 3개의 `BehaviorRelay`를 두어 사용했습니다.
<br>

```swift
let todoWorks = BehaviorRelay<[Work]>(value: [])
let doingWorks = BehaviorRelay<[Work]>(value: [])
let doneWorks = BehaviorRelay<[Work]>(value: [])
```
<br>

하지만 삭제, 수정을 구현하면서 상태에 따라 사용할 `BehaviorRelay`를 선택해야 하면서 코드가 복잡해졌습니다. 특히 상태 변경을 할 때 특정 relay에서 다른 relay로 값을 넘겨주는게 비효율적이었습니다. 
<br>

```swift
let works: BehaviorRelay<[Work]>

switch work.state {
case .todo:
    works = todoWorks
case .doing:
    works = doingWorks
case .done:
    works = doneWorks
}

works.map { //... 
```
<br>

이렇게 매번 상태를 확인해 할당하는 코드를 반복적으로 사용해야 했고, 사용할 때마다 이렇게 상태 별로 분류해야 한다면 미리 나눠서 가지고 있는 의미가 없다 생각해 모든 할 일을 가지는 `BehaviorRelay`를 하나로 두고 사용할 때 `filter`로 분리해 사용하도록 변경했습니다.
<br>

```swift
// WorkViewModel.swift 
let works = BehaviorRelay<[Work]>(value: [])
```

```swift
// 사용할 때
viewModel.works
    .map {
        $0.filter { $0.state == .doing }
    }.// 생략
```
<br>

모두 문제 없이 적용되었지만 테이블뷰에서 인덱스만 알고 있을 때 해당 cell의 `work` 값을 바로 구하는 데에 어려움이 있어 `WorkViewModel`에 인덱스 번호와 상태를 통해 `work`를 반환하는 `selectWork(by index: Int, _ state: WorkState) -> Work` 메서드를 구현했습니다.

<br>

### 2. cell을 길게 누르는 제스쳐
cell을 길게 누르면 popover가 뜨도록 하기 위해 cell에 `UILongPressGestureRecognizer`를 추가했습니다. 처음에는 제스쳐 하나를 선언해 모든 cell에 추가해주었더니 제대로 동작하지 않는 문제가 있었습니다.

공식문서를 살펴보니 recognizer는 특정 뷰와 그 하위 뷰에만 적용이 가능함을 알 수 있었습니다. 
> "A gesture recognizer operates on touches hit-tested to a specific view and all of that view’s subviews.""

<br> 따라서 cell 별로 다른 제스처 recognizer를 가져야 하는데 매번 생성하는 코드를 작성하는 것은 비효율적이라 아래와 같이 클로저 형식으로 구현해 상태를 받아 새로운 `UILongPressGestureRecognizer`를 반환하도록 했습니다.

```swift
let longGesture: (WorkState) -> UILongPressGestureRecognizer = { state in
    switch state {
    case .todo:
        return UILongPressGestureRecognizer(target: self, action: #selector(self.showTodoPopView))
    case .doing:
        return UILongPressGestureRecognizer(target: self, action: #selector(self.showDoingPopView))
    case .done:
        return UILongPressGestureRecognizer(target: self, action: #selector(self.showDonePopView))
    }
}
// ...

cell.addGestureRecognizer(longGesture(item.state))
```
<br>

## 궁금한 점 💭
### 재진입 문제
할 일을 수정하는 코드를 작성하면서 콘솔창에 아래와 같은 경고가 발생했습니다.
<br>

![](https://i.imgur.com/YUa1Xsx.png)
<br>

스케줄을 `observe(on: MainScheduler.instance)`로 설정했을 때 이와 같은 경고가 발생하였고, 경고의 내용으로는 재진입과 관련된 문제로 만약 의도한 동작일 경우 `observe(on: ainScheduler.asyncInstance)`를 통해 경고를 해결할 수 있다고 하여 변경하니 경고는 더 이상 뜨지 않았습니다.
<br>

```swift
let works = BehaviorRelay<[Work]>(value: [])

func editWork(_ work: Work, newWork: Work) {
    works.map {
        $0.map {
            return $0.id == work.id ? newWork : $0
        }
    }.observe(on: MainScheduler.asyncInstance)
    .take(1)
    .bind(to: works)
    .disposed(by: disposeBag)
}
```
<br>

재진입하는 동작을 비동기로 돌려 처리하는 것 같은데, 왜 재진입하는 문제가 발생하는 것이고 이렇게 처리해도 괜찮은 것인지 궁금합니다..!
<br>

그리고 같은 문제인데 해결되지 않은 경우도 있습니다 🥲 cell을 꾹 눌러 popover를 통해 일의 상태를 변경하는 버튼에 `tap`으로 동작을 연결했습니다. 버튼을 눌러 상태를 변경해 다른 테이블뷰로 이동하는 동작은 잘 작동했는데 똑같은 경고가 발생했습니다. 하지만 `asyncInstance`으로 변경해도 경고가 사라지지 않아 아래처럼 다양한 경우로 실험해봤습니다.
<br>

```swift
// ChangeWorkStateViewController.swift
moveTodoButton.rx.tap	// 1) 스레드를 지정하지 않은 경우
    .subscribe(onNext: {
        viewModel.chnageWorkState(work, to: .todo)
        self.dismiss(animated: true)
    }).disposed(by: disposeBag)

moveDoingButton.rx.tap	
    .observe(on: MainScheduler.asyncInstance) // 2) asyncInstance
    .subscribe(onNext: {
        viewModel.chnageWorkState(work, to: .doing)
        self.dismiss(animated: true)
    }).disposed(by: disposeBag)

moveDoneButton.rx.tap	
    .observe(on: MainScheduler.instance) // 3) instance
    .subscribe(onNext: {
        viewModel.chnageWorkState(work, to: .done)
        self.dismiss(animated: true)
    }).disposed(by: disposeBag)
```
<br>

> 실행 결과
> 1) 스레드를 지정하지 않은 경우 -> 변경되지만 경고 발생
> 2) asyncInstance -> 변경되지만 경고 발생
> 3) instance -> 경고는 발생하지 않지만 힐 일이 사라질 뿐 이동되지 않음

<br>

동작은 구현할 수 있었지만 경고를 해결하려면 이 경우에는 어떤 이유에서 문제가 발생했고 또 어떻게 해결해야 할까요? 😭

Judy-999 committed:
refactor: 날짜 지역화를 preferredLanguages 사용하는 방식으로 변경

- 헤더뷰의 backgroundColor 변경

Judy-999 committed:
feat: 새로운 할 일 추가 구현

Judy-999 committed:
feat: 할 일을 수정할 때와 추가할 때의 뷰를 다르게 구성

- 현재 수정인지 추가인지 나타내는 ViewMode 열거형 추가
- 수정일 때 edit 버튼을 누르기 전엔 수정할 수 없도록 적용

Judy-999 committed:
feat: Work에 id와 state 항목 추가 및 WorkState 타입 생성

- id: 할 일을 구분하는 고유번호
- state: 할 일의 현재 상태를 나타내는 열거형 값

Judy-999 committed:
feat: 할 일 수정 기능 구현

Judy-999 committed:
feat: 할 일을 스와이프로 삭제하는 기능 구현

Judy-999 committed:
chore: Mark 주석 수정

Judy-999 committed:
feat: 할 일의 상태 변경 popover 뷰를 표시할 뷰컨트롤러 구현

- ChangeWorkStateViewController 생성
- work에 따라 버튼을 구성하도록 구현

Judy-999 committed:
feat: 할 일(cell)을 길게 누르면 상태 변경 popover 뜨도록 구현

Judy-999 committed:
refactor: WorkViewModel이 하나의 할 일 리스트만 가지도록 변경

Judy-999 committed:
feat: 다른 상태로 이동하는 기능 구현

Judy-999 committed:
refactor: 할 일 상태 변경하는 로직 수정

- Work 수정하는 메서드를 WorkViewModel로 이동

Judy-999 committed:
refactor: 테이블뷰의 헤더를 section 헤더로 적용해 스크롤 시 고정되도록 변경

Judy-999 committed:
refactor: 상태 변경 버튼의 tap 스레드 명시

wonhee009 commented on ProjectManager/ProjectManager/View/ChangeWorkStateViewController.swift:
```diff
@@ -0,0 +1,97 @@
+    }
+    
+    // MARK: - Methods
+    func setupView(for work: Work, _ viewModel: WorkViewModel) {
+        switch work.state {
+        case .todo:
+            verticalStackView.addArrangedSubview(moveDoingButton)
+            verticalStackView.addArrangedSubview(moveDoneButton)
+        case .doing:
+            verticalStackView.addArrangedSubview(moveTodoButton)
+            verticalStackView.addArrangedSubview(moveDoneButton)
+        case .done:
+            verticalStackView.addArrangedSubview(moveTodoButton)
+            verticalStackView.addArrangedSubview(moveDoingButton)
+        }
+        
+        bindButtonTap(work: work, viewModel: viewModel)
+    }
```
해당 VC는 롱탭제스처시 뜨는 팝오버VC로 이해했는데요.
[이렇게](https://www.ioscreator.com/tutorials/display-popover-ios-tutorial) 모달로 띄울 수 있지 않나요?

입력받은 work.state에 따라 3개의 버튼 중 2개의 버튼을 insert하지 않고 가능할까요?
어떤 work.state가 들어와도 버튼은 2개일텐데 버튼 label만 바꿔줘도 되지 않나요?

또한, work 상태를 바꿔주기 위해 viewModel을 받는건 이해했지만 어색한거 같습니다.
viewModel을 받지 않고 할 수 있나요?

Judy-999 replied:
올려주신 링크는 스토리보드에 정의된 뷰컨트롤러를 이용한다는 차이점을 제외하고 비슷한 코드인 것 같은데 혹시 어떤 방법을 말씀해주신 걸까요..? 🥲

Judy-999 replied:
새로운 뷰컨트롤러를 사용하는 것이 번거롭고 `viewModel`을 넘겨주지 않고 처리할 방법을 찾아보다,

<img width="538" alt="스크린샷 2022-09-22 오전 10 37 46" src="https://user-images.githubusercontent.com/102353787/191639413-91f3f344-24c2-4a70-8cbd-eaf350aebcd3.png">

요구사항에 액션시트를 사용해도 된다는 글을 보고 생성한 뷰컨트롤러 대신에 `UIAlertController`를 사용하니 동일한 UI popover를 띄우는게 가능하여 변경했습니다!

혹시 원래 라자냐가 의도하셨던 방향은 어떤 방식이었을까요..?🤔

wonhee009 replied:
앗 제가 생각한 부분이 UIAlertController 사용하는 방식였는데요. 제가 게시글을 착각했나봅니다 ㅠ_ㅠ

wonhee009 commented on ProjectManager/ProjectManager/View/HeaderView.swift:
```diff
@@ -39,8 +39,8 @@ final class HeaderView: UIView {
-        self.backgroundColor = .systemBackground
-        self.frame = CGRect(x: 0, y: 0, width: 0, height: 50)
+        self.backgroundColor = .systemGray6
+//        self.frame = CGRect(x: 0, y: 0, width: 0, height: 50)
```
안쓰는 코드는 지워주세요.

Judy-999 replied:
삭제했습니다..!

wonhee009 commented on ProjectManager/ProjectManager/View/ManageWorkViewController.swift:
```diff
@@ -6,9 +6,22 @@
+    // MARK: - Properties
+    private let disposeBag = DisposeBag()
     private let workManageView = WorkManageView()
+    private var viewMode: ViewMode = .edit
```
뷰모드의 초기 상태가 edit인가요?

Judy-999 replied:
기본 상태를 두었는데 수정과 추가가 모두 가능한 뷰인데 기본을 두는 것은 의미에 맞지 않는 것 같아 옵셔널 상태로 주고 이후 변경해서 사용하도록 수정했습니다!

wonhee009 commented on ProjectManager/ProjectManager/View/ManageWorkViewController.swift:
```diff
@@ -17,28 +30,71 @@ final class ManageWorkViewController: UIViewController {
+    }
+    
+    @objc private func doneBarButtonTapped() {
+        guard let viewModel = viewModel else { return }
```
viewModel가 Optional인가요?

Judy-999 replied:
`viewModel`을 외부에서 주입받아 사용하고 있어서 

`private var viewModel: WorkViewModel?`

현재 이렇게 옵셔널타입이라 guard로 바인딩해준 후 사용했습니다!

wonhee009 replied:
아핫 제가 viewMode를 보고 non-Optional로 생각했네요.
viewModel을 Optional로 두기 보다는
```swift
class AViewController {
   private let viewModel: AViewModel
   init(viewModel: AViewModel = AViewModel()) {
      self.viewModel = viewModel
   }
}
```
이렇게 init 시점에서 주입받는건 어떤가요?

wonhee009 commented on ProjectManager/ProjectManager/View/ManageWorkViewController.swift:
```diff
@@ -17,28 +30,71 @@ final class ManageWorkViewController: UIViewController {
+        case .add:
+            guard let newWork = workManageView.createNewWork() else { return }
+
+            viewModel.works.accept([newWork] + viewModel.works.value)
```
viewModel의 works는 VM에서도 읽고 쓸 수 있고, VC에서도 읽고 쓸 수 있는데요.
쓰기는 VM에서만 가능하도록하고 VC에서는 읽기만 하는건 어떤가요?

```swift
class viewModel{
   private let aSubject<>
   let aObservable<>
}
```

Judy-999 replied:
```swift
private let works = BehaviorSubject<[Work]>(value: [])
let worksObservable: Observable<[Work]>
```

외부에선 `worksObservable`만 사용하도록 했습니다 :)

wonhee009 commented on ProjectManager/ProjectManager/View/ManageWorkViewController.swift:
```diff
@@ -17,28 +30,71 @@ final class ManageWorkViewController: UIViewController {
+        viewMode = .add
+    }
+    
+    func configureEditMode(with work: Work, _ viewModel: WorkViewModel) {
         workManageView.configure(with: work)
+        workManageView.changeEditMode(false)
+        self.viewModel = viewModel
+        self.work = work
     }
```
여기서는 viewMode을 바꿔줄 필요 없나요?

Judy-999 replied:
여기에서도 `edit`모드로 전환하도록 `self.viewMode = .edit` 코드를 추가했습니다.

wonhee009 reviewed (changes requested):
수고하셨습니다.
코멘트 남겨두었으니 확인 부탁드려요~

[이 답변](https://stackoverflow.com/questions/58332584/rxswift-mainscheduler-instance-vs-mainscheduler-asyncinstance)이 도움이 될거 같습니다 :)

Judy-999 committed:
refactor: 수정 및 추가하는 뷰컨에서 기본 ViewMode 지정하지 않고 사용

- 사용하지 않는 주석 제거

Judy-999 committed:
refactor: 새로운 work를 추가하는 코드를 뷰모델로 이동

- addWork 메서드로 구현

Judy-999 committed:
refactor: 읽기만 가능한 worksObservable 생성

- 기존의 work는 BehaviorSubject<[Work]>로 변경하고 private으로 설정

Judy-999 committed:
refactor: work를 가져오고 삭제할 때 id를 이용하도록 변경

- cell에 work와 동일한 id를 부여

Judy-999 committed:
refactor: 할 일 상태 변경을 AlertController를 사용하는 방식으로 변경

- ChangeWorkStateViewContoller 대신 UIAlertContoroller 사용

Judy-999 committed:
refactor: 할 일 상태를 변경하는 메서드의 로직 변경

- 변경 전: 원래 할 일 삭제 + 변경된 할 일 추가
- 변경 후: 원래 할 일 대신 변경된 할 일을 삽입
- 스레드 관련 경고 해결

Judy-999 committed:
chore: 사용하지 않는 뷰컨트롤러 제거

- ChangeWorkStateViewController 삭제
- ProjectManagerViewController 메서드 extension으로 분리

wonhee009 reviewed (approved):
수고하셨습니다 :)
남겨둔 코멘트는 다음 스텝에서 생각해주세요
